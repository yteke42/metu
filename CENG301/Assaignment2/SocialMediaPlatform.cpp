#include "SocialMediaPlatform.h"
#include "User.h"
#include "SortedLinkedList.h"
#include "LinkedListNode.h"

using namespace std;

// Implement functions in the "SocialMediaPlatform.h"
SocialMediaPlatform::SocialMediaPlatform() : users(new SortedLinkedList()) {}

SocialMediaPlatform::~SocialMediaPlatform() { delete users; }

bool SocialMediaPlatform::addUser(const int userId, const std::string name) {
    if (users->checkIfExist(userId)) return false;

    User *newUser = new User();

    newUser->setId(userId);
    newUser->setName(name);
    users->add(newUser);
    return true;
} 

bool SocialMediaPlatform::removeUser(const int userId) {
    // Iterate through all users in the platform
    LinkedListNode *currentUserNode = users->getHead();
    while (currentUserNode != nullptr) {
        // Remove the user to be deleted from the current user's followeds and followers list
        currentUserNode->u->removeFollowed(userId);
        currentUserNode->u->removeFollower(userId);

        // Move to the next user
        currentUserNode = currentUserNode->next;
    }

    // Remove the user from the platform
    return users->remove(userId);
}





bool SocialMediaPlatform::followUser(const int firstUserId, const int secondUserId) {
    LinkedListNode *firstUserNode = users->getPointerTo(firstUserId);
    LinkedListNode *secondUserNode = users->getPointerTo(secondUserId);

    if (firstUserNode == nullptr || secondUserNode == nullptr) return false;

    if (firstUserNode->u->getFolloweds()->checkIfExist(secondUserId)) {
        // User is already following the second user
        return false;
    }

    firstUserNode->u->addFollowed(secondUserNode->u);
    secondUserNode->u->addFollower(firstUserNode->u);
    return true;
}


bool SocialMediaPlatform::unfollowUser(const int firstUserId, const int secondUserId) {
    LinkedListNode *firstUserNode = users->getPointerTo(firstUserId);
    LinkedListNode *secondUserNode = users->getPointerTo(secondUserId);

    if (firstUserNode == nullptr || secondUserNode == nullptr) return false;

    if (!firstUserNode->u->getFolloweds()->remove(secondUserId) ||
        !secondUserNode->u->getFollowers()->remove(firstUserId)) {
        return false; // Either user not found or user not followed, return false
    }

    return true;
}


void SocialMediaPlatform::showFollowersOf(const int userId) const {
    LinkedListNode *userNode = users->getPointerTo(userId);
    if (userNode != nullptr) {
        std::cout << "Users following the user with ID " << userId << ":" << std::endl;
        if (!userNode->u->getFollowers()->isEmpty()) {
            userNode->u->getFollowers()->print();
        } else {
            std::cout << "None" << std::endl;
        }
    } else {
        std::cout << "Cannot show. There is no user with ID " << userId << "." << std::endl;
    }
}





void SocialMediaPlatform::showAllUsers() const {
    std::cout << "Users in the social media platform:" << std::endl;
    if (users->isEmpty()) {
        std::cout << "None" << std::endl;
    } else {
        users->print();
    }
}





