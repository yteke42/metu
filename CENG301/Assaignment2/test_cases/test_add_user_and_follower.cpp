#include <iostream>
using namespace std;

#include "SocialMediaPlatform.h"

int main() {

    string truth_table[2] = {"false", "true"};

    SocialMediaPlatform SMP;
    
    SMP.showAllUsers();
    cout<<endl;

    cout<<truth_table[SMP.addUser( 3, "Ali" )]<<endl;
    cout<<truth_table[SMP.addUser( 2, "Ayse" )]<<endl;
    cout<<truth_table[SMP.addUser( 4, "Fatma" )]<<endl;
    cout<<truth_table[SMP.addUser( 4, "Bora" )]<<endl;
    cout<<truth_table[SMP.addUser( 1, "Pelin" )]<<endl;
    
    SMP.showAllUsers();
    cout<<endl;
    
    cout<<truth_table[SMP.followUser( 1, 2)]<<endl;
    cout<<truth_table[SMP.followUser( 1, 3)]<<endl;
    cout<<truth_table[SMP.followUser( 1, 4)]<<endl;
    cout<<endl;
    
    cout<<truth_table[SMP.followUser( 2, 1)]<<endl;
    cout<<truth_table[SMP.followUser( 2, 3)]<<endl;
    cout<<truth_table[SMP.followUser( 2, 4)]<<endl;
    cout<<endl;
    
    cout<<truth_table[SMP.followUser( 3, 1)]<<endl;
    cout<<truth_table[SMP.followUser( 3, 2)]<<endl;
    cout<<truth_table[SMP.followUser( 3, 4)]<<endl;
    cout<<endl;
    
    cout<<truth_table[SMP.followUser( 4, 1)]<<endl;
    cout<<truth_table[SMP.followUser( 4, 2)]<<endl;
    cout<<truth_table[SMP.followUser( 4, 3)]<<endl;
    cout<<endl;
    
    cout<<truth_table[SMP.followUser( 1, 2)]<<endl;
    cout<<truth_table[SMP.followUser( 1, 3)]<<endl;
    cout<<truth_table[SMP.followUser( 1, 4)]<<endl;
    cout<<endl;
    
    
    SMP.showFollowersOf(1);
    cout<<endl;
    SMP.showFollowersOf(2);
    cout<<endl;
    SMP.showFollowersOf(3);
    cout<<endl;
    SMP.showFollowersOf(4);
    cout<<endl;
    SMP.showFollowersOf(4);
    cout<<endl;
    
    

    return 0;
}
