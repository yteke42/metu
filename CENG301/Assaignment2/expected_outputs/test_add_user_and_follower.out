Users in the social media platform:
None

true
true
true
false
true
Users in the social media platform:
1, Pelin
2, Ayse
3, Ali
4, Fatma

true
true
true

true
true
true

true
true
true

true
true
true

false
false
false

Users following the user with ID 1:
2, Ayse
3, Ali
4, Fatma

Users following the user with ID 2:
1, Pelin
3, Ali
4, Fatma

Users following the user with ID 3:
1, Pelin
2, Ayse
4, Fatma

Users following the user with ID 4:
1, Pelin
2, Ayse
3, Ali

Users following the user with ID 4:
1, Pelin
2, Ayse
3, Ali

